target/mips/cpu.h

```diff
@@ -1518,6 +1485,42 @@ static inline void mips_update_pc(CPUMIPSState *env, target_ulong pc_addr, bool
     mips_update_pc_impl(&env->active_tc, pc_addr, can_be_unrepresenable);
 }
 
+
+#ifdef CONFIG_TCG_LOG_INSTR
+/*
+ * Check whether the cpu is in user mode.
+ */
+static inline bool cpu_in_user_mode(CPUMIPSState *env)
+{
+    return ((env->hflags & MIPS_HFLAG_UM) == MIPS_HFLAG_UM);
+}
+
+/*
+ * Target-specific hook to fetch the cpu log state
+ */
+static inline cpu_log_instr_info_t *cpu_get_log_instr_state(CPUMIPSState *env)
```

Comment:
Should this be part of the target-indepedent data instead?

Reply:
On second thought, I think I can move it back to the `CPUState` and either have a `GArray` of log info or some other semi-opaque struct that gets allocated in `qemu_log_instr_init`. This would also align with the requirements of the ring-buffer idea.